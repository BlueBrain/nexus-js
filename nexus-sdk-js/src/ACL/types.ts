export type Context = string | string[];

export type IdentityType = 'User' | 'Group' | 'Authenticated' | 'Anonymous';

export interface Identity {
  '@id': string;
  '@type': IdentityType;
  subject?: string;
  realm?: string;
  group?: string;
}

export interface ACLResponseCommon {
  '@id': string;
  '@type': IdentityType;
  _path: string;
  acl: {
    permissions: string[];
    identity: Identity;
  }[];
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
}

export interface ACLResponse extends ACLResponseCommon {
  '@context': Context;
}

export interface ListACLResponse {
  '@context': Context;
  _total: number;
  _results: ACLResponseCommon[];
}

export interface ACLOperationSuccess {
  '@context': Context;
  '@id': string;
  '@type': 'AccessControlList';
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
}

export interface ListRealmOption {
  from?: number;
  size?: number;
  rev?: number;
  ancestors?: true;
  self?: boolean;
  [key: string]: any;
}

export interface ACLPayload {
  permissions: string[];
  identity: {
    subject?: string;
    realm?: string;
    group?: string;
  };
}
