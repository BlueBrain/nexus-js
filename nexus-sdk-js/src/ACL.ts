export type PermissionsEndpointResponse = {
  "@context"?: string | string[];
  "@id": string;
  "@type": string;
  permissions: string[];
  _rev: number;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
}

export type IdentityType = 'User' | 'Group' | 'Authenticated' | 'Anonymous';

export type IdentitiesEndpointResponse = {
  '@context'?: string | string[];
  identities: IdentityResponse[];
}

export type IdentityResponse = {
  '@id': string;
  '@type': IdentityType;
  subject?: string;
  realm?: string;
  group?: string;
};

export type ACLResponse = {
  path: string;
  identity: IdentityResponse,
  permissions: Permission[],
};

export class Identity {
  id: string;
  type: IdentityType;
  realm?: string;
  group?: string;
  sub?: string;

  constructor(identityResponse: IdentityResponse) {
    this.id = identityResponse['@id'];
    this.type = identityResponse['@type'];
  }
}

export default class ACL {
  path: string;
  permissions: Permission[];
  identity: Identity;

  constructor(aclResponse: ACLResponse) {
    this.path = aclResponse.path;
    this.identity = new Identity(aclResponse.identity);
    this.permissions = aclResponse.permissions;
  }
}
