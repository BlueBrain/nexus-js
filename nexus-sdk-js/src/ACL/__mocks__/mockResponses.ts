import { ACLResponse, ListACLResponse } from '../types';

export const mockACLResponse: ACLResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/resource.json',
    'https://bluebrain.github.io/nexus/contexts/acls.json',
  ],
  '@id': 'https://nexus.example.com/v1/acls/org1',
  '@type': 'User',
  _path: '/org1',
  acl: [
    {
      permissions: ['read'],
      identity: {
        '@id': 'https://nexus.example.com/v1/realms/myrealm/groups/a-group',
        '@type': 'Group',
        realm: 'myrealm',
        group: 'a-group',
      },
    },
    {
      permissions: ['read', 'write'],
      identity: {
        '@id': 'https://nexus.example.com/v1/realms/myrealm/groups/some-group',
        '@type': 'Group',
        realm: 'myrealm',
        group: 'some-group',
      },
    },
    {
      permissions: ['acls/read', 'acls/write'],
      identity: {
        '@id': 'https://nexus.example.com/v1/realms/myrealm/users/user',
        '@type': 'User',
        realm: 'myrealm',
        subject: 'alice',
      },
    },
  ],
  _createdAt: '2018-09-18T09:58:00.801Z',
  _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _updatedAt: '2018-09-18T10:01:00.801Z',
  _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _rev: 1,
};

export const mockListACLResponse: ListACLResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/resource.json',
    'https://bluebrain.github.io/nexus/contexts/iam.json',
    'https://bluebrain.github.io/nexus/contexts/search.json',
  ],
  _total: 2,
  _results: [
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj',
      '@type': 'Authenticated',
      acl: [
        {
          permissions: ['read', 'write'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realm/groups/two',
            '@type': 'Group',
            realm: 'myrealm',
            group: 'two',
          },
        },
      ],
      _path: '/myorg/myproj',
      _rev: 1,
      _createdAt: '2018-09-17T14:55:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
      _updatedAt: '2018-09-17T15:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
    },
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj2',
      '@type': 'Anonymous',
      acl: [
        {
          permissions: ['read'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realms/myrealm/users/me',
            '@type': 'User',
            realm: 'myrealm',
            subject: 'me',
          },
        },
      ],
      _path: '/myorg/myproj2',
      _rev: 2,
      _createdAt: '2018-09-17T14:00:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
      _updatedAt: '2018-09-17T14:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
    },
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj',
      '@type': 'Authenticated',
      acl: [
        {
          permissions: ['read', 'write'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realm/groups/two',
            '@type': 'Group',
            realm: 'myrealm',
            group: 'two',
          },
        },
      ],
      _path: '/myorg/myproj',
      _rev: 1,
      _createdAt: '2018-09-17T14:55:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
      _updatedAt: '2018-09-17T15:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
    },
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj2',
      '@type': 'Anonymous',
      acl: [
        {
          permissions: ['read'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realms/myrealm/users/me',
            '@type': 'User',
            realm: 'myrealm',
            subject: 'me',
          },
        },
      ],
      _path: '/myorg/myproj2',
      _rev: 2,
      _createdAt: '2018-09-17T14:00:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
      _updatedAt: '2018-09-17T14:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
    },
  ],
};
