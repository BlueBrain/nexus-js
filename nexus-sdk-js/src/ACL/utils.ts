import ACL from '.';
import { ACLResponse } from './types';
import { httpGet } from '../utils/http';

export async function getACL(
  path: string,
  rev?: number,
  self?: boolean,
): Promise<ACL> {
  try {
    const aclResponse: ACLResponse = await httpGet(`/acls/${path}`);
    return new ACL(aclResponse);
  } catch (error) {
    throw error;
  }
}
