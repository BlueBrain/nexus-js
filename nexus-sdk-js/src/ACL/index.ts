import { IdentityType, Identity, ACLResponse } from './types';
import { getACL, listACL } from './utils';

export default class ACL {
  id: string;
  type: IdentityType;
  path: string;
  acl: {
    permissions: string[];
    identity: Identity;
  }[];
  createdAt: string;
  createdBy: string;
  updatedAt: string;
  updatedBy: string;
  rev: number;

  static get = getACL;
  static list = listACL;

  constructor(aclResponse: ACLResponse) {
    this.id = aclResponse['@id'];
    this.type = aclResponse['@type'];
    this.path = aclResponse._path;
    this.acl = aclResponse.acl;
    this.createdAt = aclResponse._createdAt;
    this.createdBy = aclResponse._createdBy;
    this.updatedAt = aclResponse._updatedBy;
    this.updatedBy = aclResponse._updatedBy;
    this.rev = aclResponse._rev;
  }
}
