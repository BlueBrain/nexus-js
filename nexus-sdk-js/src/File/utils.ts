import { FileResponse } from './types';
import { httpPost, HttpConfigTypes } from '../utils/http';
import { File as NexusFile } from '..';

export async function createFile(
  orgLabel: string,
  projectLabel: string,
  file: File,
): Promise<NexusFile> {
  try {
    const fileResponse: FileResponse = await httpPost<File>(
      `/files/${orgLabel}/${projectLabel}`,
      file,
      { sendAs: HttpConfigTypes.FILE },
    );
    return new NexusFile(orgLabel, projectLabel, fileResponse);
  } catch (error) {
    throw new Error(error);
  }
}
