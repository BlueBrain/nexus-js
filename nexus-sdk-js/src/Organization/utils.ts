import { Resource, Organization } from '..';
import { OrgResponse, ListOrgsResponse, ListOrgsOptions } from '.';
import { httpPut, httpGet, httpDelete } from '../utils/http';
import { CreateOrganizationException } from './exceptions';

/**
 *
 * @param label The label of your organization
 * @param name The name of your organization
 */
export async function createOrganization(
  label: string,
  name: string,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(`/orgs/${label}`, {
      name,
    });
    return new Organization({
      ...orgResponse,
      projectNumber: 0,
      _deprecated: false,
    });
  } catch (error) {
    throw new CreateOrganizationException(error.message);
  }
}

// TODO: refactor -> blocked by https://github.com/BlueBrain/nexus/issues/112
/**
 *
 * @param orgLabel The organization label to fetch
 * @param Object<revision, tag> The specific tag OR revision to fetch
 */
export async function getOrganization(
  orgLabel: string,
  options?: { revision?: number; tag?: string },
): Promise<Organization> {
  try {
    // check if we have options
    let ops = '';
    if (options) {
      // it's rev or tag, not both. We take rev over tag
      if (options.revision) {
        ops = `?rev=${options.revision}`;
      } else if (options.tag) {
        ops = `?tag=${options.tag}`;
      }
    }
    const orgResponse: OrgResponse = await httpGet(`/orgs/${orgLabel}${ops}`);
    // we want to know how many projects there are per organisation
    const listOrgsResponse: ListOrgsResponse = await httpGet('/projects');
    const projectNumber: number = listOrgsResponse._results
      ? listOrgsResponse._results.reduce((prev, org) => {
          const split = org._id.split('/');
          const orgName = split.slice(split.length - 2, split.length - 1)[0];
          if (orgName === orgLabel) {
            return prev + 1;
          }
          return prev;
        }, 0)
      : 0;

    const org = new Organization({ ...orgResponse, projectNumber });
    return org;
  } catch (e) {
    throw new Error(`ListOrgsError: ${e}`);
  }
}

// TODO: refactor -> blocked by https://github.com/BlueBrain/nexus/issues/112
// cannot implement list orgs options until then...
export async function listOrganizations(
  options?: ListOrgsOptions,
): Promise<Organization[]> {
  let ops;
  if (options) {
    ops = Object.keys(options).reduce(
      (currentOps, key) =>
        currentOps.length === 0
          ? `?${key}=${options[key]}`
          : `${currentOps}&${key}=${options[key]}`,
      '',
    );
  }
  try {
    const listOrgsResponse: ListOrgsResponse = await httpGet('/projects');
    if (listOrgsResponse.code || !listOrgsResponse._results) {
      return [];
    }
    // Get list of unique orgs names
    const filteredOrgNames: string[] = listOrgsResponse._results
      .map(org => {
        const split = org._id.split('/');
        const orgName = split.slice(split.length - 2, split.length - 1)[0];
        return orgName;
      })
      .filter((org, index, self) => self.indexOf(org) === index);

    // get orgs details
    return Promise.all(
      filteredOrgNames.map(async org => await getOrganization(org)),
    );
  } catch (e) {
    throw new Error(`ListOrgsError: ${e}`);
  }
}

/**
 *
 * @param orgLabel Current organization label
 * @param rev Last known revision
 * @param newName Knew name the organization will be called with
 */
export async function updateOrganization(
  orgLabel: string,
  rev: number = 1,
  newName: string,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${orgLabel}?rev=${rev}`,
      {
        name: newName,
      },
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 *
 * @param orgLabel Current organization label
 * @param rev Last know revision
 * @param Object<{tagName, revision}> The name of the tag and revision number to tag the organization from
 */
export async function tagOrganization(
  orgLabel: string,
  rev: number = 1,
  {
    tagName,
    tagFromRev,
  }: {
    tagName: string;
    tagFromRev: number;
  },
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${orgLabel}/tags?rev=${rev}`,
      {
        tag: tagName,
        rev: tagFromRev,
      },
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 *
 * @param orgLabel Label of the Org to be deprecated
 * @param rev Last know revision
 */
export async function deprecateOrganization(
  orgLabel: string,
  rev: number = 1,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpDelete(
      `/orgs/${orgLabel}?rev=${rev}`,
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}
