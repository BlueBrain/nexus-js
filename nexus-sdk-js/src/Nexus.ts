import Store from './utils/Store';
import Organization, { OrganizationResponse } from './Organization';
import { httpGet, httpPost } from './utils/http';

type NexusConfig = {
  environment: string;
  token?: string;
};

export const store: Store = new Store({
  auth: {
    accessToken: null,
  },
  api: {
    baseUrl: null,
  },
});

export default class Nexus {
  constructor(config: NexusConfig) {
    if (!config.environment) {
      throw new Error(
        'No environment provided. Please specify your Nexus instance endpoint.'
      );
    }
    store.update('api', state => ({
      ...state,
      baseUrl: config.environment,
    }));
    if (config.token) {
      store.update('auth', state => ({
        ...state,
        accessToken: config.token,
      }));
    }
  }

  setToken(token: string): void {
    store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  removeToken(): void {
    store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  async listOrganizations(): Promise<Organization[]> {
    try {
      const orgs: OrganizationResponse[] = await httpGet('/orgs');
      return orgs.map((org: OrganizationResponse) => new Organization(org));
    } catch (e) {
      console.log(e);
      return e;
    }
  }

  async getOrganization(id: string): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpGet(`/orgs/${id}`);
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }

  async createOrganization(): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpPost('/orgs', {});
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }
}
