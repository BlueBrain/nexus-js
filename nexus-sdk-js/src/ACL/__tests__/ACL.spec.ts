import Nexus from '../../Nexus';
import { resetMocks, mockResponse, mock } from 'jest-fetch-mock';
import { mockACLResponse } from '../__mocks__/mockResponses';
import ACL from '..';
import { getACL } from '../utils';

const baseUrl = 'http://api.url';
Nexus.setEnvironment(baseUrl);

describe('ACL utils', () => {
  describe('ACL class', () => {
    it('should return an instance of ACL', () => {
      const acl = new ACL(mockACLResponse);
      expect(acl).toBeInstanceOf(ACL);
    });
    it('should match snapshot', async () => {
      const acl = new ACL(mockACLResponse);
      expect(acl).toMatchSnapshot();
    });
  });

  describe('utils()', () => {
    describe('getAcl()', () => {
      beforeEach(() => {
        mockResponse('{}');
      });
      afterEach(() => {
        resetMocks();
      });
      it('should make a GET request', async () => {
        await getACL('myorg');
        expect(mock.calls[0][0]).toEqual(`${baseUrl}/acls/myorg`);
      });
      it('should make a GET request with rev query param', async () => {
        await getACL('myorg', { rev: 12 });
        expect(mock.calls[0][0]).toEqual(`${baseUrl}/acls/myorg?rev=12`);
      });
      it('should make a GET request with self query param', async () => {
        await getACL('myorg', { self: true });
        expect(mock.calls[0][0]).toEqual(`${baseUrl}/acls/myorg?self=true`);
      });
      it('should make a GET request with all query params', async () => {
        await getACL('myorg', { rev: 34, self: true });
        expect(mock.calls[0][0]).toEqual(
          `${baseUrl}/acls/myorg?rev=34&self=true`,
        );
      });
    });
  });
});
