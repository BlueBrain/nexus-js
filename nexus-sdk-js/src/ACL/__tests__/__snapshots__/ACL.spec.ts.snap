// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ACL utils ACL class should match snapshot 1`] = `
ACL {
  "acl": Array [
    Object {
      "identity": Object {
        "@id": "https://nexus.example.com/v1/realms/myrealm/groups/a-group",
        "@type": "Group",
        "group": "a-group",
        "realm": "myrealm",
      },
      "permissions": Array [
        "read",
      ],
    },
    Object {
      "identity": Object {
        "@id": "https://nexus.example.com/v1/realms/myrealm/groups/some-group",
        "@type": "Group",
        "group": "some-group",
        "realm": "myrealm",
      },
      "permissions": Array [
        "read",
        "write",
      ],
    },
    Object {
      "identity": Object {
        "@id": "https://nexus.example.com/v1/realms/myrealm/users/user",
        "@type": "User",
        "realm": "myrealm",
        "subject": "alice",
      },
      "permissions": Array [
        "acls/read",
        "acls/write",
      ],
    },
  ],
  "createdAt": "2018-09-18T09:58:00.801Z",
  "createdBy": "https://nexus.example.com/v1/realms/myrealm/users/john",
  "id": "https://nexus.example.com/v1/acls/org1",
  "path": "/org1",
  "rev": 1,
  "type": "User",
  "updatedAt": "https://nexus.example.com/v1/realms/myrealm/users/john",
  "updatedBy": "https://nexus.example.com/v1/realms/myrealm/users/john",
}
`;
